bitcoin_client_js/src/lib/psbtv2.ts

```diff
@@ -371,6 +376,88 @@ export class PsbtV2 {
       while (this.readKeyPair(this.outputMaps[i], buf));
     }
   }
+  /**
+   * Imports a BitcoinJS (bitcoinjs-lib) Psbt object.
+   * https://github.com/bitcoinjs/bitcoinjs-lib/blob/master/ts_src/psbt.ts
+   *
+   * Prepares the fields required for signing a Psbt on a Ledger
+   * device. It should be used exclusively before calling 
+   * `appClient.signPsbt()` and not as a general Psbt conversion method.
+   *
+   * Note: This method supports all the policies that the Ledger is able to
+   * sign, with the exception of taproot: tr(@0).
+   */
+  fromBitcoinJS(psbtBJS: bjs.Psbt) : PsbtV2 {
+    function isTaprootInput(input): boolean {
+      let isP2TR;
+      try {
+        bjs.payments.p2tr({ output: input.witnessUtxo.script });
+        isP2TR = true;
+      } catch (err) {
+        isP2TR = false;
+      }
+      return (
+        input &&
+        !!(
+          input.tapInternalKey ||
+          input.tapMerkleRoot ||
+          (input.tapLeafScript && input.tapLeafScript.length) ||
+          (input.tapBip32Derivation && input.tapBip32Derivation.length) ||
+          isP2TR
+        )
+      );
+    }
+    this.setGlobalPsbtVersion(2);
+    this.setGlobalTxVersion(psbtBJS.version);
+    this.setGlobalInputCount(psbtBJS.data.inputs.length);
+    this.setGlobalOutputCount(psbtBJS.txOutputs.length);
+    if (psbtBJS.locktime !== undefined)
+      this.setGlobalFallbackLocktime(psbtBJS.locktime);
+    psbtBJS.data.inputs.forEach((input, index) => {
+      if (isTaprootInput(input))
+        throw new Error(`Taproot inputs not supported`);
+      this.setInputPreviousTxId(index, psbtBJS.txInputs[index].hash);
+      if (psbtBJS.txInputs[index].sequence !== undefined)
+        this.setInputSequence(index, psbtBJS.txInputs[index].sequence);
+      this.setInputOutputIndex(index, psbtBJS.txInputs[index].index);
+      if (input.sighashType !== undefined)
+        this.setInputSighashType(index, input.sighashType);
+      if (input.nonWitnessUtxo)
+        this.setInputNonWitnessUtxo(index, input.nonWitnessUtxo);
+      if (input.witnessUtxo) {
+        this.setInputWitnessUtxo(
+          index,
+          input.witnessUtxo.value,
+          input.witnessUtxo.script
+        );
+      }
+      if (input.witnessScript)
+        this.setInputWitnessScript(index, input.witnessScript);
+      if (input.redeemScript)
+        this.setInputRedeemScript(index, input.redeemScript);
+      psbtBJS.data.inputs[index].bip32Derivation.forEach(derivation => {
+        if (derivation.path.substring(0, 2) !== 'm/')
+          throw new Error(`Invalid input bip32 derivation`);
+        const pathArray = derivation.path
+          .replace('m/', '')
+          .split('/')
+          .map(level =>
+            level.match("'") ? parseInt(level) + 0x80000000 : Number(level)
```

Comment:
Does BitcoinJS always use the `'` symbol for hardened derivation? (that is: it's never `h`)?
Otherwise, `.match("['h]")` handles both; or maybe:
```TypeScript
parseInt(level) + (['h', "'"].includes(level.slice(-1)) ? 0x80000000 : 0)
```

Reply:
That should be fine. Take a look at the implementation of the "bip32" module in bitcoinjs-lib:

https://github.com/bitcoinjs/bip32/blob/master/ts-src/bip32.ts#L384
Note that bitcoinjs-lib uses a lowercase "m" to denote the master key, and single quotes for hardened paths.

Thank you for accepting the PR, Salvatore. Next week, I'll be releasing a TypeScript descriptors module that will benefit from this change.

Reply:
In any case I'll go ahead and implement those changes to ensure that it won't break if bitcoinjs-lib changes in the future. Better be safe than sorry.